### PR TITLE
Implements retry logic for sending events

### PR DESCRIPTION
**Pull request by puckpuck**
Adds a ring buffer with configurable size and expiration timeout to be leveraged by the libhoney response channel. As events are sent to Honeycomb, they are added to the buffer. If during response handling, the event was not successfully sent, the event will be retried for sending.  Libhoney is also now configured to block when the libhoney buffers are full.

The buffer size and timeout are configurable.  Setting the buffer size to 0, will disable the usage of the buffer, and continue operation as before this change.

The buffer will most certainly increase memory consumption, especially on startup when the agent is backfilling logs. The buffer should never be set to more than 16000 (2 x libhoney buffer of 8000). Even a buffer of 16000 can increase initial memory consumption by 150+MB.  As such a recommended default of 1000 which provides retry protection should Honeycomb service return a 500 status, with a modest memory consumption increase.

### CODE REVIEW

File: examples/quickstart.yaml [hunk 1]
```diff
@@ -46,6 +46,8 @@ metadata:
 data:
   config.yaml: |
     apiHost: https://api.honeycomb.io/
+    retryBufferSize: 1000
```
Comment (JamieDanielson) [inline examples/quickstart.yaml:49]: Should we default to 0 to keep existing behavior, or is the retry logic critical enough that it should be used? I guess when we add it to the helm chart we could put in these default values but commented out...

Comment (puckpuck) [reply examples/quickstart.yaml:49]: Yeah, we can set this to 0 for the default, and document it with recommended values, etc.

File: docs/configuration-reference.md [hunk 2]
```diff
@@ -323,6 +323,19 @@ additionalFields:
...
+The default size is 0 which means retry is not enabled.
+The recommendation when first enabling retry is to set this buffer size to 1000 which provides some buffer for temporary Honeycomb API ingest errors with only a modest increase in memory consumption by the agent.
+Note: there is no flushing of the events in the retry buffer on process shutdown.
+
+### retryBufferExpire
+
+When enabling retry, you may opt to expire events out of the retry buffer.
+The default expiration is 0 which means events will live on the retry buffer for the lifetime of the running agent process.
+If you wish to expire events from the buffer, set this to the number of seconds an individual event may be considered for retry before removed from the buffer unsent.
+
```
Comment (robbkidd) [inline docs/configuration-reference.md:338]: @puckpuck I attempted some wordsmithing in the config reference about these new options. Can you check whether they make sense?

Comment (MikeGoldsmith) [reply docs/configuration-reference.md:338]: We're using a `time.Duration` type to capture the event expiry, so it should be expressed as a value with time suffix. eg `"500ms"`, `"10s"`, `"1m"`, etc

File: transmission/transmission.go [hunk 3]
```diff
@@ -2,80 +2,157 @@ package transmission
...
+				if exists {
+					// event found, we can retry it
+					logrus.WithFields(logrus.Fields{
+						"error":        resp.Err,
+						"status":       resp.StatusCode,
+						"responseBody": string(resp.Body),
+					}).Error("Failed to send event to Honeycomb. Will retry.")
+
+					// Resend event
+					ht := &HoneycombTransmitter{}
```
Comment (MikeGoldsmith) [inline transmission/transmission.go:137]: Do we need to create a new transmitter for each retry attempt?

Comment (puckpuck) [reply transmission/transmission.go:137]: We need to put the event back on the ring buffer to retry, so we need to call `Send` on a transmitter. HoneycombTrasmistter is an empty struct, so this shouldn't have any impact.

File: docs/configuration-reference.md (hunk 2 above)
Comment (puckpuck) [reply docs/configuration-reference.md:340]: Clarified the docs, and added the time suffix portion

**Pull request merged**